ast/src/main/java/com/graphicsfuzz/common/tool/PrettyPrinterVisitor.java

```diff
@@ -572,10 +572,34 @@ public void visitTernaryExpr(TernaryExpr ternaryExpr) {
 
   @Override
   public void visitArrayIndexExpr(ArrayIndexExpr arrayIndexExpr) {
-    visit(arrayIndexExpr.getArray());
-    out.append("[");
-    visit(arrayIndexExpr.getIndex());
-    out.append("]");
+    /* If arrayIndexExpr indexes into a literal-to-uniform array, output a #define instead. */
+    if (arrayIndexExpr.getArray().getText().contains(Constants.INT_LITERAL_UNIFORM_VALUES)
+        && uniformValues.isPresent()
+        && uniformValues.get().getValues(Constants.INT_LITERAL_UNIFORM_VALUES).isPresent()) {
+      int index = ((IntConstantExpr) arrayIndexExpr.getIndex()).getNumericValue();
+      List<String> values =
+          uniformValues.get().getValues(Constants.INT_LITERAL_UNIFORM_VALUES).get();
+      out.append("_int_" + values.get(index));
+    } else if (arrayIndexExpr.getArray().getText().contains(Constants.UINT_LITERAL_UNIFORM_VALUES)
+        && uniformValues.isPresent()
+        && uniformValues.get().getValues(Constants.UINT_LITERAL_UNIFORM_VALUES).isPresent()) {
+      int index = ((IntConstantExpr) arrayIndexExpr.getIndex()).getNumericValue();
+      List<String> values =
+          uniformValues.get().getValues(Constants.UINT_LITERAL_UNIFORM_VALUES).get();
+      out.append("_uint_" + values.get(index));
+    } else if (arrayIndexExpr.getArray().getText().contains(Constants.FLOAT_LITERAL_UNIFORM_VALUES)
+        && uniformValues.isPresent()
+        && uniformValues.get().getValues(Constants.FLOAT_LITERAL_UNIFORM_VALUES).isPresent()) {
+      int index = ((IntConstantExpr) arrayIndexExpr.getIndex()).getNumericValue();
+      List<String> values =
+          uniformValues.get().getValues(Constants.FLOAT_LITERAL_UNIFORM_VALUES).get();
+      out.append("_float_" + values.get(index).replace(".", "_"));
+    } else {
+      visit(arrayIndexExpr.getArray());
+      out.append("[");
+      visit(arrayIndexExpr.getIndex());
+      out.append("]");
+    }
```

Comment:
Is there a way to avoid all this duplication?

How about a function:

```
private bool maybeEmitKnownUniformMacro(ArrayIndexExpr arrayIndexExpr, String knownUniformArrayName, String prefix);
```

This function could check whether the array index expr's array matches the array name, in which case it could emit the symbol and return true.  Otherwise it could return false.

Then you could do something like:

```
if (maybeEmitKnownUnformMacro(arrayIndexExpr, Constants.INT_LITERAL_UNIFORM_VALUES, "int") {
  return;
}
if (maybeEmitKnownUnformMacro(arrayIndexExpr, Constants.UINT_LITERAL_UNIFORM_VALUES, "uint") {
  return;
}
...
// The original code
```

Reply:
I added the method and cleaned up the duplication.